### PR TITLE
packaging: Overhaul Containerfile and use Temurin JRE

### PR DESCRIPTION
**Pull request by k3rnelpan1c-dev**
### Description

I props these changes as I stumbled over this great tool while dealing with the aftermath that was the Log4Shell hell week and noticed that the Dockerfile could be improved. I did so since we would deploy dtrack to our OpenShift cluster which uses arbitrary UIDs, which is currently already supported by the fronted container yet not by the API server. And while I have made similar changes to packaging for an internal draft container to test deploy dtrack I wanted to offer these changes upstream as well.

Additionally, I noticed that there is the motivation to use the Eclipse-Temurin JVM based on the comment in https://github.com/DependencyTrack/dependency-track/issues/1213#issuecomment-978807113, so I took the liberty and migrated the Container to that alongside the other changes.

To answer questions as to why I did so via a [multi stage build](https://docs.docker.com/develop/develop-images/multistage-build/) I advice to read the following to issues that may be a long read but should explain it:

* https://github.com/adoptium/containers/issues/1 (also covers the glibc issue with Alpine and announces that there will be a Musl build of the Temurin JVM eventually :tada: )
* https://github.com/adoptium/containers/issues/54

***TLDR***: DockerHub apparently does not accept `slim` images anymore for their `Official Image` labelled images and the 'old' [`adoptopenjdk`](https://hub.docker.com/_/adoptopenjdk) too did only offer Ubuntu, CentOS7 and WindowsServerCore images. So the common consensus is to utilize multi stage builds to copy the JDK/JRE from a official image to a derived image with the same package manager (i.e. Ubuntu to Debian, CentOS to UBI or Fedora, etc.), a good example could be the [Jenkins Official Docker images](https://github.com/jenkinsci/docker/blob/master/11/debian/bullseye-slim/hotspot/Dockerfile), which use Jlink to minify the JDK and then copy it to their image. (I did the same without Jlink since the JRE does not have that nor need it to be slim)

If there are any further questions or any feedback please feel free to share them I will come back to them as soon as possible.

### Changes

* overhauls and (IMHO) simplifies/groups Dockerfile
* migrates Dockerfile to Eclipse-Temurin from Zulu JRE
* update `maven` workflow slightly 

### Issues

* relates to https://github.com/DependencyTrack/dependency-track/issues/1213#issuecomment-978807113
* relates to https://github.com/DependencyTrack/dependency-track/issues/1090
* similar to https://github.com/DependencyTrack/dependency-track/pull/994
* similar to https://github.com/DependencyTrack/dependency-track/pull/891

### Notes

I noticed that you seem to use a bash script to release new version of dtrack and want to propose moving that to CI (GitHub Actions) as well. I would be up for submitting the necessary PRs to do so if you are interested :slightly_smiling_face: 

**EDIT:** there are now official musl builds of Temurin 11, 16 and 17 available, so if you ever want to offer an alpine spin alongside the debian one you can now without any glibc hackery for the Alpine JVM variant.

### CODE REVIEW

File: .github/workflows/maven.yml [hunk 1]
```diff
@@ -12,66 +12,79 @@ on:
...
-    - name: Install tools
-      run: |
-        sudo apt-get install jq
-        wget -O ~/codacy-coverage-reporter-assembly.jar https://github.com/codacy/codacy-coverage-reporter/releases/download/4.0.5/codacy-coverage-reporter-4.0.5-assembly.jar
-        export DOCKLE_VERSION=$(curl --silent "https://api.github.com/repos/goodwithtech/dockle/releases/latest" | grep '"tag_name":' | sed -E 's/.*"v([^"]+)".*/\1/')
-        wget -O ~/dockle.tar.gz https://github.com/goodwithtech/dockle/releases/download/v${DOCKLE_VERSION}/dockle_${DOCKLE_VERSION}_Linux-64bit.tar.gz
-        tar zxvf ~/dockle.tar.gz -C ~/
-        export TRIVY_VERSION=$(curl --silent "https://api.github.com/repos/aquasecurity/trivy/releases/latest" | grep '"tag_name":' | sed -E 's/.*"v([^"]+)".*/\1/')
-        wget -O ~/trivy.tar.gz https://github.com/aquasecurity/trivy/releases/download/v${TRIVY_VERSION}/trivy_${TRIVY_VERSION}_Linux-64bit.tar.gz
-        tar zxvf ~/trivy.tar.gz -C ~/
```
Comment (k3rnelpan1c-dev) [inline .github/workflows/maven.yml:48]: JQ and Docker are both already installed in a recent version in the GitHub Actions Virtual environment and since you don't seem to use `trivy`, additionally it having an official action otherwise, I cleaned up this overhead

**Comment by k3rnelpan1c-dev** [comment]
@stevespringett sorry for the ping, but could I get your feedback on this and whether you would be interested in a GitHub Action based release workflow instead or as an alternative to your `release.sh`?

**Comment by stevespringett** [comment]
Thanks for the PR. I'll need a few days to review.

File: src/main/docker/Dockerfile [hunk 2]
```diff
@@ -1,63 +1,70 @@
-FROM debian:stable-slim
-LABEL maintainer="steve.springett@owasp.org"
-LABEL vendor="OWASP"
+FROM eclipse-temurin:11.0.13_8-jre-focal AS jre-build
```
Comment (nscuro) [inline src/main/docker/Dockerfile:1]: We should go ahead and jump to `17.0.2_8-jre-focal` directly.

File: src/main/docker/Dockerfile [hunk 3]
```diff
@@ -1,63 +1,70 @@
...
 
 # Arguments that can be passed at build time
 # Directory names must end with / to avoid errors when ADDing and COPYing
+ARG COMMIT_SHA
+ARG APP_VERSION
 ARG APP_DIR=/opt/owasp/dependency-track/
 ARG DATA_DIR=/data/
-ARG USERNAME=dtrack
+ARG UID=1001
+ARG GID=1001
```
Comment (nscuro) [inline src/main/docker/Dockerfile:12]: Is there a specific reason why `1001` was chosen? The previous image used `1000`, so it's safe to assume that the permissions of most mounted volumes in current deployments are geared towards it (as in `1000:1000`). 

If we keep `1001` as the default, it should be documented in the upgrade notes. But as always, there will be quite a few users running into issues when blindly upgrading.

**Review by nscuro — COMMENTED** [review]
LGTM overall. Have a few minor suggestions, and the `UID` situation needs clarification.

File: src/main/docker/Dockerfile (hunk 3 above)
Comment (k3rnelpan1c-dev) [reply src/main/docker/Dockerfile:12]: ah my bad, I likely used the UID i used on another Image of mine I sourced for the changes. I will revert to `1000:1000`.
Thank you for that find!

File: src/main/docker/Dockerfile [hunk 4]
```diff
@@ -1,63 +1,70 @@
-FROM debian:stable-slim
-LABEL maintainer="steve.springett@owasp.org"
-LABEL vendor="OWASP"
+FROM eclipse-temurin:11.0.13_8-jre-focal AS jre-build
+
+FROM debian:bullseye-slim
```
[the remark this answers is not in the document]
Comment (k3rnelpan1c-dev) [reply src/main/docker/Dockerfile:3]: Will do, should have done that to begin with :sweat_smile:

File: src/main/docker/Dockerfile (hunk 2 above)
Comment (k3rnelpan1c-dev) [reply src/main/docker/Dockerfile:1]: Oh okay, so directly to the new LTS? 
I have heard of ppl running into issues at work who compiled against 11 and ran on 17.
Just want to throw out this as a note, but I will change it if you still want me to.

Comment (stevespringett) [reply src/main/docker/Dockerfile]: @k3rnelpan1c-dev I've been testing with Java 17, but the testing has not been extensive. Lets hold off on it for now. I'm concerned about some of the changes to reflection.

**Comment by namloc2001** [comment]
@k3rnelpan1c-dev - looks really good, the only thing that is missing for full OpenShift compatibility is:
```
# Specify this so that when OpenShift restricted SCC assigns arbitrary UID - it is aware of the correct home dir to use
ENV HOME ${DATA_DIR}
```

This is what I use in my Dockerfile that uses `FROM dependencytrack/apiserver:4.3.6` as base image. (DATA_DIR is used because that is the home directory assigned to dtrack user and the chown & chmod statements make it accessible when OpenShift assigns an arbitrary UID and GID=0 to the container user under the restricted SCC).

Are you please able to add it in?

@Homopatrol FYI.

**Comment by k3rnelpan1c-dev** [comment]
Hi,

first of all thank you two for this observation!
I work with OpenShift at work as well and have run/tested this changes there before I contributed them and never noticed that the user home was even listed in the logs.
However, mind me asking if any of you happen to know if this is a big issue? 
I have only a limited set of test and evaluation experience with Dependency Track on OpenShift, but as far as I was able to tell it never complained nor ran into issues caused by the improperly set home dir. The likely reason for that is that Dependency Track stores all of its data to the `/data/` dir and largely, if not entirely, ignoring the home dir despise the log output by Alpine.

**Comment by namloc2001** [comment]
@k3rnelpan1c-dev it appears that the container can certainly start up and analyse an SBOM with the home directory defaulting to `/`, however I would suggest that best practice is followed for compatibility and the environment variable is added in.

Also, on a separate note to this topic, I've noticed you've amended the `HEALTHCHECK` statement. I believe that by removing the "/" but not updating the comment associated with the `CONTEXT` environment variable, you risk someone setting one without a trailing slash and breaking the healthcheck command.

**Comment by k3rnelpan1c-dev** [comment]
Hey @namloc2001,
I am not opposed to adding the envVar, was just wondering :wink: 
Regarding the Context, that you are indeed right that it could mess with the Healthcheck :man_facepalming:.
Unfortunately, I messed actually messed the Healtcheck up with this PR which then needed to be fixed in a followup PR #1408 and that is now in 4.4.1 (yay for the OCI standard not supporting `HEALTHCHECK` and me not testing the Docker format xD).

All things considered, I will offer the HOME env addition in a future PR and have to check the Healthcheck again :/
Thank you for the comments and the pointer for yet another issue with the Health check :expressionless:
